cpp/src/arrow/compute/kernels/scalar_nested.cc

```diff
@@ -428,40 +428,82 @@ const FunctionDoc make_struct_doc{"Wrap Arrays into a StructArray",
                                    "specified through MakeStructOptions."),
                                   {"*args"},
                                   "MakeStructOptions"};
-
+template <typename KeyType>
 struct MapArrayLookupFunctor {
-  static Result<int64_t> FindOneMapValueIndex(const Array& keys, const Scalar& query_key,
+  static Result<int64_t> FindOneMapValueIndex(const Array& keys,
+                                              const Scalar& query_key_scalar,
                                               const int64_t start, const int64_t end,
                                               const bool from_back = false) {
-    if (!from_back) {
-      for (int64_t idx = start; idx < end; ++idx) {
-        ARROW_ASSIGN_OR_RAISE(std::shared_ptr<Scalar> key, keys.GetScalar(idx));
-
-        if (key->Equals(query_key)) return idx;
-      }
-    } else {
-      for (int64_t idx = end - 1; idx >= start; --idx) {
-        ARROW_ASSIGN_OR_RAISE(std::shared_ptr<Scalar> key, keys.GetScalar(idx));
+    const auto query_key = UnboxScalar<KeyType>::Unbox(query_key_scalar);
+    int64_t index = 0;
+    int64_t match_idx = -1;
+    ARROW_UNUSED(VisitArrayValuesInline<KeyType>(
+        *keys.data(),
+        [&](decltype(query_key) key) -> Status {
+          if (index < start) {
+            ++index;
+            return Status::OK();
+          } else if (index < end) {
+            if (key == query_key) {
+              if (!from_back) {
+                match_idx = index;
+                return Status::Cancelled("Found first matching key");
+              } else {
+                match_idx = index;
+              }
+            }
+            ++index;
+            return Status::OK();
+          } else {
+            return Status::Cancelled("End reached");
+          }
+        },
+        [&]() -> Status {
+          if (index < end) {
+            ++index;
+            return Status::OK();
+          } else {
+            return Status::Cancelled("End reached");
+          }
+        }));
 
-        if (key->Equals(query_key)) return idx;
-      }
-    }
-    return -1;
+    return match_idx;
   }
 
   static Result<std::unique_ptr<ArrayBuilder>> GetBuiltArray(
```

Comment:
This could perhaps take in `ArrayBuilder* builder` as a parameter instead of returning it. That way, for the Array case below, it could directly append to the child items builder of the list builder, avoiding an extra copy of  the data/allocation of a scalar. And the Scalar case could just use it as normal.